kyuubi-main/src/test/scala/org/apache/kyuubi/server/MiniYarnService.scala

```diff
@@ -0,0 +1,106 @@
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.server
+
+import java.io.File
+import java.net.InetAddress
+import java.util.concurrent.TimeUnit
+
+import scala.collection.JavaConverters._
+
+import org.apache.hadoop.security.UserGroupInformation
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+import org.apache.hadoop.yarn.server.MiniYARNCluster
+
+import org.apache.kyuubi.Utils
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.service.AbstractService
+
+class MiniYarnService(name: String) extends AbstractService(name) {
+  def this() = this(classOf[MiniYarnService].getSimpleName)
+
+  private var hadoopConfDir: File = _
+  private var yarnConf: YarnConfiguration = _
+  private var yarnCluster: MiniYARNCluster = _
+
+  private def newYarnConfig(): YarnConfiguration = {
+    val yarnConfig = new YarnConfiguration()
+    // Disable the disk utilization check to avoid the test hanging when people's disks are
+    // getting full.
+    yarnConfig.set("yarn.nodemanager.disk-health-checker.max-disk-utilization-per-disk-percentage",
+      "100.0")
+
+    // capacity-scheduler.xml is missing in hadoop-client-minicluster so this is a workaround
+    yarnConfig.set("yarn.scheduler.capacity.root.queues", "default")
+    yarnConfig.setInt("yarn.scheduler.capacity.root.default.capacity", 100)
+    yarnConfig.setFloat("yarn.scheduler.capacity.root.default.user-limit-factor", 1)
+    yarnConfig.setInt("yarn.scheduler.capacity.root.default.maximum-capacity", 100)
+    yarnConfig.set("yarn.scheduler.capacity.root.default.state", "RUNNING")
+    yarnConfig.set("yarn.scheduler.capacity.root.default.acl_submit_applications", "*")
+    yarnConfig.set("yarn.scheduler.capacity.root.default.acl_administer_queue", "*")
+    yarnConfig.setInt("yarn.scheduler.capacity.node-locality-delay", -1)
+
+    // Set bind host to localhost to avoid java.net.BindException
+    yarnConfig.set("yarn.resourcemanager.bind-host", "localhost")
+
+    // enable proxy
+    val currentUser = UserGroupInformation.getCurrentUser.getShortUserName
+    yarnConfig.set(s"hadoop.proxyuser.$currentUser.groups", "*")
+    yarnConfig.set(s"hadoop.proxyuser.$currentUser.hosts", "*")
+    yarnConfig
+  }
+
+  override def initialize(conf: KyuubiConf): Unit = {
+    hadoopConfDir = Utils.createTempDir().toFile
+    yarnConf = newYarnConfig()
+    yarnCluster = new MiniYARNCluster(name, 1, 1, 1)
+    yarnCluster.init(yarnConf)
+    super.initialize(conf)
+  }
+
+  override def start(): Unit = {
+    yarnCluster.start()
+    val config = yarnCluster.getConfig()
+    val startTimeNs = System.nanoTime()
+    while (config.get(YarnConfiguration.RM_ADDRESS).split(":")(1) == "0") {
+      if (System.nanoTime() - startTimeNs > TimeUnit.SECONDS.toNanos(10)) {
+        throw new IllegalStateException("Timed out waiting for RM to come up.")
+      }
+      debug("RM address still not set in configuration, waiting...")
+      TimeUnit.MILLISECONDS.sleep(100)
+    }
```

Comment:
eventually -> config.get(YarnConfiguration.RM_ADDRESS).split(":")(1) == "0"?

Reply:
Will refactor it.